### DISCUSSION TITLE
Proposed GraphQL changes in NetBox v4.0

### DISCUSSION THREAD

**Discussion by arthanson** [netbox-community/netbox#13583] [General]
A heads up for people using GraphQL, as referenced in issue #9856 we are looking at replacing the current GraphQL library with a new one called Strawberry.  A side effect of this is that the query syntax would change so I wanted to give people a heads up and open it up for discussion.  The difference comes in the filtering syntax as shown below:

| Old                                                   | New                                                                              |
|----------------------------------|---------------------------------------------------|
| user_list(is_superuser: true) {      | user_list(filters: {is_superuser: true}) {                     |
| user_list(username__ic: "bob") { | user_list(filters: {username: {i_contains: "bob"}}) { |
| user_list(id: "3") {                          | user_list(filters: {id: {exact: 3}}) {                               |

So you have to add "filters:" to the query and the lookup syntax for case-insensitive comparisons and less-the, greater-then queries also changes to be a bit more verbose. Or, more explicitly:

**Graphene**
```
user_list(is_superuser: true) {
    id
}
```
**Strawberry**
```
user_list(filters: {is_superuser: true}) {
   id
}
```
The filtering lookup syntax (greater than, less than, case insensitive comparison) is also different:
**Graphene**
```
{
  user_list(username__ic: "bob") {
    username
  }
}
```
**Strawberry**
```
{
  user_list(filters: {username: {i_contains: "bob"}}) {
    username
  }
}
```

For stawberry, allowing lookups on fields forces the lookup to be present, so it forces this somewhat awkward syntax:
**Graphene**
```
{
  user_list(id: "3") {
    id
  }
}
```
**Strawberry**
```
{
  user_list(filters: {id: {exact: 3}}) {
    id
  }
}
```

**Comment by jsenecal** [comment]
I only have a few scripts (one-liners) using graphql. I think this is not much of an issue _for me_.

I would however strongly advise documenting the lookup names as they are not the same as those from the REST api.

**Comment by glesys-andreas** [comment]
Sounds good! I don't think it will be any major issues. Also I agree with jsenecal that if you deviate from the REST api naming that would be reflected in the documentation.
